crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+//! State relating to validating a WebAssembly component.
+
+use super::{check_max, core::Module};
+use crate::{
+    limits::*,
+    validator::core::{EntityType, ModuleTypeSpace},
+    BinaryReaderError, CanonicalOption, ExternalKind, FuncType, GlobalType, MemoryType, Result,
+    TableType, Type, WasmFeatures,
+};
+use std::{collections::HashMap, rc::Rc, sync::Arc};
+
+fn check_options(options: &[CanonicalOption], offset: usize) -> Result<()> {
```

Comment:
In the same manner that core wasm modules have a `ValidatorResources` or w/e trait I think we'll want to expose type information from the validator for components as well. For functions I think we'll want to expose which option was selected, if any, for `encoding` and `into`.

Reply:
Agreed, I think there should be a follow-up PR to export this internal information from the validator in a way that makes sense for other libraries to consume.